test/org/zaproxy/zap/extension/pscanrulesAlpha/UserControlledCharsetScannerUnitTest.java

```diff
@@ -0,0 +1,191 @@
+		return new UserControlledCharsetScanner();
+	}
+
+	public HttpMessage createMessage() {
+		HttpMessage msg = new HttpMessage();
+		HttpRequestHeader requestHeader = new HttpRequestHeader();
+		try {
+			requestHeader.setURI(new URI("http://example.com/i.php", false));
+		} catch (URIException | NullPointerException e) {
+		}
+		requestHeader.setMethod(HttpRequestHeader.GET);
+
+		msg = new HttpMessage();
+		msg.setRequestHeader(requestHeader);
+		msg.getResponseHeader().setStatusCode(HttpStatusCode.OK);
+		msg.getResponseHeader().addHeader(HttpHeader.CONTENT_TYPE, "text/html");
+		return msg;
+	}
+
+	@Test
+	public void shouldNotRaiseAlertIfResponseIsNotStatusOk() {
+		// Given
+		HttpMessage msg = createMessage();
+		msg.getResponseHeader().setStatusCode(HttpStatusCode.NOT_ACCEPTABLE);
+		// When
+		rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+		// Then
+		assertThat(alertsRaised.size(), equalTo(0));
+	}
+
+	@Test
+	public void shouldNotRaiseAlertIfRequestHasNoGetParams() {
+		// Given
+		HttpMessage msg = createMessage();
+		// WHen
+		rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+		// Then
+		assertThat(alertsRaised.size(), equalTo(0));
+	}
+
+	@Test
+	public void shouldNotRaiseAlertIfResponseIsNotHtml() {
+		// Given
+		HttpMessage msg = createMessage();
+		msg.getResponseHeader().setHeader(HttpHeader.CONTENT_TYPE, "application/json");
+		// When
+		rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+		// Then
+		assertThat(alertsRaised.size(), equalTo(0));
+	}
+
+	@Test
+	public void shouldNotRaiseAlertIfRequestParamsHaveNoValues() throws URIException, NullPointerException {
+		// Given
+		HttpMessage msg = createMessage();
+		msg.getRequestHeader().setURI(new URI("http://example.com/i.php?place=&name=", false));
+		msg.getResponseHeader().setHeader(HttpResponseHeader.SET_COOKIE, "Set-Cookie: aCookie=aValue; Secure");
+		// When
+		rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+		// Then
+		assertThat(alertsRaised.size(), equalTo(0));
+	}
+
+	@Test
+	public void shouldNotRaiseAlertIfResponseCharsetIsEmpty() throws URIException, NullPointerException {
+		// Given
+		HttpMessage msg = createMessage();
+		msg.getRequestHeader().setURI(new URI("http://example.com/i.php?cs=utf-8", false));
+		msg.getResponseHeader().setHeader(HttpResponseHeader.CONTENT_TYPE, "text/html; charset=");
+		// When
+		rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+		// Then
+		assertThat(alertsRaised.size(), equalTo(0));
+	}
+
+	@Test
+	public void shouldRaiseAlertIfRequestParamsAppearAsCharsetValue() throws URIException, NullPointerException {
+		// Given
+		HttpMessage msg = createMessage();
+		msg.getRequestHeader().setURI(new URI("http://example.com/i.php?cs=utf-8", false));
+		msg.getResponseHeader().setHeader(HttpResponseHeader.CONTENT_TYPE, "text/html; charset=utf-8");
+		// When
+		rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+		// Then
+		assertThat(alertsRaised.size(), equalTo(1));

```

Comment:
Could also assert the alert's param (same for following tests).

Reply:
Yup for the `should` set agreed.